dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/reservedvalue/RandomPatternBuilder.java

```diff
@@ -0,0 +1,258 @@
+ * and/or other materials provided with the distribution.
+ * Neither the name of the HISP project nor the names of its contributors may
+ * be used to endorse or promote products derived from this software without
+ * specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package org.hisp.dhis.reservedvalue;
+
+import java.math.BigInteger;
+import java.security.SecureRandom;
+import java.util.List;
+import java.util.Objects;
+import java.util.UUID;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+import java.util.stream.Stream;
+
+import org.apache.commons.lang3.StringUtils;
+
+/**
+ * @author Luca Cambi <luca@dhis2.org>
+ *
+ *         Pattern builder class. Provides string builders used to append
+ *         segments of a random pattern.
+ *
+ *         For alphanumeric, we have some corner cases:
+ *
+ *         - '*' : it shuffles between lower and upper case and gets a number
+ *         when it finds it in UUID random string
+ *
+ *         - 'x' or 'X' : in UUID random string gets the letter ( and make it
+ *         upper case for 'X' ) but if we find a number, it gets the alphabetic
+ *         char at that position in order to maintain random pattern
+ *
+ *         For numeric patterns that are longer than 1 digit, we also have
+ *         builders that contains all zeros or partial zeros beacuase random
+ *         digits are generated from BigInteger that has no leading zeros, so we
+ *         need to consider zero leading cases.
+ *
+ *         For digits:
+ *
+ *         - '#': in case of at least 2 digits pattern inside the segment, we
+ *         also create a second dimension that contains leading zeros ( numbers
+ *         of zero is randomly determined ) and a third dimension whit only
+ *         zeros. If the segment has only digits, we generate an all zero
+ *         pattern only once, otherwise we append to it for only iteration.
+ *
+ *         In case of 1 digit, we just generate random numbers in range 0 - 9
+ *
+ */
+public class RandomPatternBuilder
+{
+    private final List<Character> lowercase = IntStream.range( 0, 26 ).mapToObj( n -> (char) (n + 'a') )
```

Comment:
Maybe something like this is a little bit cleaner:
`IntStream.rangeClosed('a, 'z').mapToObj(c -> (char) c).collect(Collectors.toList())`